torch_glow/src/FusingOptimizer.cpp

```diff
@@ -68,4 +73,106 @@ graph(%input, %weight):
 }
 
+const torch::jit::Symbol& getGlowSymbol() {
+  static torch::jit::Symbol glow_sym = torch::jit::Symbol:fromQualString("glow::CompilationGroup");
+  return glow_symbol;
+}
+
+value_list sortReverseTopological(ArrayRef<Value*> inputs, Block* block) {
+  value_list result;
+  for (auto i : inputs) {
+    if (i->node()->owningBlock() == block) {
+	  result.push_back(i);
+	}
+  }
+
+  std::sort(result.begin(), result.end(), [&](Value* a, Value* b) {
+    return a->node()->isAfter(b->node());
+	});
+  return result;
+}
+
+bool canMerge(jit::torch::Node* node, isSupportFunc fn) {
+  return node->kind == torch::jit::prim::Constant || fn(node);
+}
+
+bool canMerge(jit::torch::Block* block, isSupportFunc fn) {
+  for (torch::jit::Node* node : block->nodes()) {
+    if (!canMerge(node, fn) {
+	  rerturn false;
+	}
+  }
+  return true;
+}
+
+#define REQ(cond)							\
+  if (!(cond)) {							\
+    llvm:err("Failed cond " #cond "\n");	\
+    return c10:nullptr;						\
+  }
+
+torch::c10::optional<torch::jit::Node*> tryMerge(torch::jit::Node* consumer, torch::jit::Node* producer, AliasDb& aliasdb, isSupportFunc fn) {
+
+  REQ(canMerge(producer, fn));
+  REQ(consumer->kind() == getGlowSymbol() || canMerge(consumer, aliasDb));
+
+  // Alias checks
+  // Requirement:
+  // - moveAfterTopologicallyValid(consumer, producer)
+  // - One of:
+  //   1) Both are in-place ops
+  //   2) Consumer is in-place, producer !hasInputWriters
+  //   3) Producer is in-place, consumer !hasOutputWriters
+  REQ(aliasDb.moveAfterTopologicallyValid(consumer, producer));
+
+  // 1)
+  if (!(aliasDb.isMutable(consumer) && aliasDb.isMutable(producer))) {
+	  // 2)
+	  if (aliasDb.isMutable(consumer)) {
+		  REQ(!aliasDb.hasInputWriters(producer));
+	  // 3)
+	  } else if (aliasDb.isMutable(producer)) {
+		  REQ(!aliasDb.hasOutputWriters(consumer));
+	  }
+  }
+
+  if (!consumer->hasAttribute(attr::Subgraph) &&
+		 consumer->kind() != getGlowSymbol()) {
+    consumer = SubgraphUtils::createSingletonSubgraph(consumer, getGlowSymbol());
+  }
+ if (producer->kind() == prim::Constant) {
+   auto& subgraph = consumer->g(attr::Subgraph);
+   torch::jit::Node* in_const = subgraph->createClone(producer, [](torch::jit::Value*) -> torch::jit::Value* {
+      throw std::runtime_error("unexpected input");
+	});
+   subgraph->insertNode(in_const);
+ } else {
+   SubgraphUtils::mergeNodeIntoSubgraph(producer, consumer);
+ }
+
+ return consumer;
+}
+#undef REQ
+
+torch::jit::graph_node_list::iterator getNewNode(torch::jit::Node* node, AliasDb& aliasDb, torch::jit::Block block, isSupportFunc fn) {
+  auto node_inputs = sortReverseTopological(node->inputs(), block);
+  for (auto input : node_inputs) {
+    if (auto group = tryMerge(node, input->node(), aliasDb, fn)) {
+	  return group.value()->reverseIterator();
+	}
+  }
+  return ++node->reverseIterator();
+}
+
+void GlowCustomFuse(std::shared_ptr<Graph> graph, isSupportFunc fn, Symbol kind, size_t arg_limit) {
```

Comment:
Let's put this new stuff in a different file for clarity

Reply:
Okay I will do it once it works in one file.